tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc

```diff
@@ -126,6 +193,73 @@ __global__ void UnsortedSegmentCustomKernel(
   }
 }
 
+// SetSegmentFreq counts the frequency of occurrence of each element in
```

Comment:
I had a hard time following this function.  Have you considered phrasing `SetSegmentFreq` as a segment sum with input as an array of ones?  You could specialize the segment sum CUDA kernel to do this without having to create a temporary array.

If that idea does not work for some reason, please add a high level overview of the implementation.

Also see top level comment.